keps/sig-cloud-provider/20190704-instance-host-labelling.md

```diff
@@ -0,0 +1,174 @@
+
+### Goals
+
+- Introduce a well known label to identify the instance host of a node.
+- Define a default value for the label so that components maintain their current behavior if they use the label.
+
+### Non-Goals
+
+- Add a cloud provider API to get instance host ID.
+- Populate the instance host label with the instance host ID.
+- Controller behavior (e.g. deployment, statefulset) that considers instance host label.
+- Any decision-making that is based on the physical host label.
+
+## Proposal
+
+Add a well-known label for instance host: `topology.kubernetes.io/instance-host` and make its default value empty.
+
+The label will be named `failure-domain.kubernetes.io/instance-host` if KEP [Promoting cloud provider labels to GA](https://github.com/kubernetes/enhancements/pull/839) doesn’t land.
+
+
+### User Stories
+
+#### Story 1
+
+As an application developer, I want pods to be scheduled in the same instance host so that networking between them is faster.
+I use `topology.kubernetes.io/instance-host` as topology key in [inter-pod affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity-beta-feature) rules.
+
+#### Story 2
+
+As an application developer, I want pods to be spread in different instance hosts so that if the host goes down, my service doesn’t have to scale from zero.
+I use `topology.kubernetes.io/instance-host` as topology key in [even pod spreading](/keps/sig-scheduling/20190221-even-pods-spreading.md) rules.
+
+### Implementation Details/Notes/Constraints
+
+- The label will initially be named `topology.beta.kubernetes.io/instance-host` before its promoted to GA.
+
+- If the label is not explicitly used in inter-pod affinity or anti-affinity, current behavior in the scheduler remains.
+
+    There is one case worth mentioning:
+    An empty topology key for `preferredDuringSchedulingIgnoredDuringExecution` in inter-pod anti-affinity is interpreted as
+    "all topologies", so the instance host should be considered as well.
+    If the label value is empty, then the anti-affinity is still ensured by the hostname.
+    Additionally, a node [should always be considered in the same topology as itself](https://groups.google.com/d/msg/kubernetes-sig-cloud-provider/32N59IYXogY/FXIUHeYWDwAJ), regardless of the labels.
+
+### Risks and Mitigations
+
+- Risk: Some or all nodes are missing a value for the instance host label and an application developer tries to use the label as a topology key.
+
+  Mitigation: This will be documented as undefined behavior.
+  
+- Risk: If a cloud provider or hypervisor performs a live migration of a VM, the host will change.
+
+  Mitigation: The label should be updated. Evicting running pods is out-of-scope for the scheduler.
+  New pods are scheduled according to the new labels.
```

Comment:
This is the biggest concern I have with this KEP. YMMV and the following comments are highly dependent on the hypervisor/virtualization platform (capabilities available/exposed/configured).

@bsalamat brought up an important point:

> We have accepted the fact that clusters are dynamic in nature and scheduling decisions may be invalidated in the future. This is an assumption that other new features should make too.

I agree with him (because fundamentally that's the current implementation of the scheduler). On the other side, from a UX/ user experience perspective this is not what is known/desired. Especially in cases where hard pod anti-affinity is needed, e.g. critical stateful services. Now, a workaround would be to use `zone` (assuming zone membership is static and won't change, which might be or not be the case...) instead of `instance` label here - and the users I've worked with do that and never reported any issues with it. 

My feeling is that a majority of users would use the instance label for resiliency, which ultimately is not guaranteed when VMs migrate, labels need to be updated and workloads have to be evicted/restarted to reconcile. So why not keep using the zones label for resiliency than? I don't want to repeat every argument of the discussion here, so pointing to the former KEP which has more details: https://github.com/kubernetes/enhancements/pull/997#pullrequestreview-231865948

Some more findings from the field (real-world on-prem deployments I was involved with): when we planned to add zones support in the vSphere cloud provider we went with a two phase approach: first, implement the basic functionality. Second, try to solve the "label drift" issue. It turned out that solving the "label drift" issues was impossible, i.e. it always had consequences (impact) to user workloads. This is not what we wanted to go through, so phase two never was implemented. Did that cause a problem? Not for our users. They leverage zones topology (again that could be arbitrary on-prem) to model resiliency constructs. Whenever needed, vSphere will be configured to never move VMs if that could cause "label drift", e.g. move VM from zone 1 to zone 2 because of cluster imbalance or host failure. I understand that this might not be possible (available) with all hypervisors - which complicates the matter unfortunately.

Also, @anfernee built a [prototype](https://github.com/vmware/vsphere-affinity-scheduling-plugin) to influence scheduling decisions honoring host information. It couldn't solve the "label drift" issue and also did not get a lot of traction because our user base was happy with the current (zones) implementation.

I am not saying to keep the status quo. But assuming that we could easily update labels and restart workloads is a weak assumption, especially for stateful production critical workloads in highly dynamic clusters. Also it might be worth pointing out that not all Kubernetes users can simply modify their workloads to comply with the Kubernetes semantics, be it because they don't have the manpower, access to source code, etc.

Reply:
@tallclair do you have any thoughts on these semantics for live migration?

Reply:
I would argue that if a VM migrates, the label MUST be updated.  Whether that ripples out into a pod reschedule is a totally different question.  We don't have a rescheduler today, but hard anti-affinity might be broken by a migration, which would effectively require either a resched or that we loosen the guarantees for affinity/anti-affinity.

Reply:
(Just highlighting that live-migrating an instance/pod, and then restarting the pod because its label changed, is equivalent to just not live-migrating it in the first place.  I advocate taking the simpler option and just not engaging with this arms race..)

Reply:
@anguslees the problem is that the hypervisor likely doesn't have a way to tell if it can migrate a VM without breaking any restrictions for the pod. 
In the re-scheduling topic, note that we face the same problem when adding new taints to a node for which running pods don't have a toleration.